docs/rfc/rfc-001-storage-engine.rst

```diff
@@ -0,0 +1,179 @@
+
+Changelog
+---------
+
+- 2021-04-19: Initial Draft (gist)
+- 2021-09-02: Migrated to RFC folder, with some updates  
+
+Abstract
+--------
+
+The aspect of Tendermint that's responsible for persistence and storage (often
+"the database" internally) represents a bottle neck in the architecture of the
+platform, that the 0.36 release presents a good opportunity to correct. The
+current storage engine layer provides a great deal of flexibility that is
+difficult for users to leverage or benefit from, while also making it harder
+for Tendermint Core developers to deliver improvements on storage engine. This
+RFC discusses the possible improvements to this layer of the system.
+
+Background
+----------
+
+Tendermint has a very thin common wrapper that makes Tendermint itself
+(largely) agnostic to the data storage layer (within the realm of the popular
+key-value/embedded databases.) This flexibility is not particularly useful:
+the benefits of a specific database engine in the context of tendermint is not
+particularly well understood, and the maintenance burden for multiple backends
+is not commensurate with the benefit provided. Additionally, because the data
+storage layer is handled generically, and most tests run with an in-memory
+framework, it's difficult to take advantage of any higher-level features of a
+database engine.
+
+Ideally, developers within Tendermint will be able to interact with persisted
+data via a single interface that can function, approximately like an object
+store, and this storage interface will be able to accommodate all existing
+persistence workloads (e.g. block storage, local peer management information
+like the "address book", crash-recovery log like the WAL.) In addition to
+providing a more ergonomic interface and new semantics, by selecting a single
+storage engine tendermint can use native durability and atomicity features of
+the storage engine and simplify its own implementations. 
+
+Data Access Patterns
+~~~~~~~~~~~~~~~~~~~~
+
+Tendermint's data access patterns have the following characteristics:
+
+- data size often exceeds memory.
+
+- data is rarely mutated after it's written for most data (e.g. blocks), but
+  small amounts of workding data is persisted by nodes and is frequently
+  mutated (e.g. peer information, validator information.)
+
+- read patterns can be quite random.
+
+- crash resistance and crash recovery, provided by write-ahead-logs (in
+  consensus, and potentially for the mempool) should allow the system to
+  resume work after an unexpected shut down.
+
+Project Goals
+~~~~~~~~~~~~~
+
+As we think about replacing the current persistence layer, we should consider
+the following high level goals: 
+
+- drop dependencies on storage engines that have a CGo dependency.
+
+- encapsulate data format and data storage from higher-level services
+  (e.g. reactors) within tendermint.
+
+- select a storage engine that does not incur any additional operational
```

Comment:
I like this idea, but it seems in tension with the previous desideratum: If we want storage to be encapsulated from the rest of the application, whether it's embedded or not becomes an implementation detail, no?

Reply:
I agree that its an implementation detail, but I think that it's a good goal to avoid increasing the operational complexity of the system.

Reply:
Well, I guess that's my point: Embedding a database is potentially quite a big dependency. It'll depend on what we actually decide to use, probably, but it could actually reduce net complexity if we allowed the database to live in an adjacent process with a pipe in between. (I realize "separate process" makes it sound like a hard ops problem, but if we assume adjacency most of the hard parts disappear).

Anyway, not advocating we _should_ do that, only that we should be open to the option.

Reply:
Today it's separate processes and a pipe, and tomorrow we have a dependency on k8s (this is a joke, mostly.) 

I think I'm open to this kind of solution, I just don't have a good idea/example of what such a system would look like, or what off-the-shelf database might exist that we could use this way.

Reply:
Well, of course we'd need to be intentional about it. 🙂 

One of the issues I've had on other projects is that storage engines often have complex dependencies, and can pull in half the world to your module file. What has worked well for me is to build a simple pipe-based communication with a separate binary that has all those dependencies, but only needs to be updated for changes to the storage engine rather than triggering dependabot every time some version bump changes one of the many minor dependencies. No need for network routing—you can run either of them as a subprocess of the other, or run them side-by-side with a Unix-domain socket.

Again, though, not advocating for that here, only that in a couple cases I've actually found this _reduced_ the complexity of building, testing, and maintaining my programs. It also means we could use things that require cgo without pulling cgo into the rest of the project. 

If we plan to pick one fairly unstructured storage engine (e.g., Badger) that is already written in Go and doesn't support complex query planning, it may be overkill.

Reply:
Using an embedded db does raise an additional question: Do we want this DB to be accessible to operators of a tendermint node? If we fully embed a DB, all queries on current tendermint state have to be performed through tendermint and all DB tuning is a concern of tendermint itself. I understand the desire to keep operational complexity low, but this does mean a good deal of the observability and tuning is a responsibility of tendermint itself. 

As nodes get larger and larger and more critical, we may worsen operational complexity by not allowing node operators to reach for any of the usual db-related pressure-release valves (sharding, running primary/secondary for improved disaster recovery,  etc.)

Reply:
That's a good question, but note that both of the database engines we've talked about here (Bolt & Badger) store data on-disk and can be accessed by other tools. (Read-only access while the "owner" process is running, but read-write when it's closed). 

This is still worth considering, but I think it's probably OK if we insist that any write access to the data store be mediated by the consensus node _while the node is running_.

Reply:
I don't think anything in this project precludes the "seperate storage and core into separate binaries" architecture, which is definitely a useful way of architecting things, but I don't think we're particularly at a point where we'd need that (e.g. the binary size is small, even with badger in it, I don't think the embedded system is particularly onerous or limiting (and is what our users expect, generally.) I operators generally expect the current kind of model, and as a chain developer, I'd be pretty skeptical of a library that required/started its own sidecar binary. Not saying we couldn't do it, but I think CGo and a big binary are easier sells (the no-cgo dependency is mostly for our own build story.)

Additionally, I must confess that I'm not particularly excited (personally) by any of the storage engines that would become available if we relaxed the CGo restriction (e.g. RocksDB, SQLite, etc.)  which might be clouding my opinion above. 

I think these are all good ideas to keep in our pockets.